### PR TITLE
[remote-shuffle]Remote shuffle manager for spark3.0

### PR DESCRIPTION
**Pull request by gczsjdy**
## What changes were proposed in this pull request?

Remote-shuffle for Spark 2.4.0 ++ batch fetch, new corruption detect, custom metrics, etc.(Spark 3.0 features)

The batch fetch feature(used in AQE) implementation will be in next PR.

## How was this patch tested?

Existing & new UTs.
200GB Repartition workload(HDFS as shuffle storage) passed on cluster.

### CODE REVIEW

**Comment by carsonwang** [comment]
@gczsjdy , what was failing the checks? can you please fix it?

**Comment by gczsjdy** [comment]
@carsonwang Because Spark 3.0 was not available previously, not it should be okay : )

**Comment by carsonwang** [comment]
Great. I will merge this first.

File: oap-shuffle/remote-shuffle/src/main/scala/org/apache/spark/util/collection/RWritablePartitionedPairCollection.scala [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+ */
+
+package org.apache.spark.util.collection
+
+import java.util.Comparator
+
+import org.apache.spark.shuffle.remote.RemoteBlockObjectWriter
+
+/**
+  * NOTE: This is to override Spark 2.4.0's WritablePartitionedIterator: Changing the writeNext
```
Comment (carsonwang) [inline oap-shuffle/remote-shuffle/src/main/scala/org/apache/spark/util/collection/RWritablePartitionedPairCollection.scala:25]: 3.0?

File: oap-shuffle/remote-shuffle/README.md [hunk 2]
```diff
@@ -1,3 +1,130 @@
...
+
+When the backend storage is HDFS, we contact http://$host:$port/conf to fetch configurations. They were not locally loaded because we assume absence of local storage.
+
+```
+    spark.shuffle.remote.hdfs.storageMasterUIPort  50070
+```
+
+### Inherited Spark Shuffle Configurations
+
+These configurations are inherited from upstream Spark, they are still supported in remote shuffle. More explanations can be found in [Spark core docs](https://spark.apache.org/docs/2.4.4/configuration.html#shuffle-behavior) and [Spark SQL docs](https://spark.apache.org/docs/2.4.4/sql-performance-tuning.html).
```
Comment (carsonwang) [inline oap-shuffle/remote-shuffle/README.md:100]: Please update 2.4 to 3.0

File: oap-shuffle/remote-shuffle/src/main/scala/org/apache/spark/util/collection/RWritablePartitionedPairCollection.scala (hunk 1 above)
Comment (gczsjdy) [reply oap-shuffle/remote-shuffle/src/main/scala/org/apache/spark/util/collection/RWritablePartitionedPairCollection.scala]: Done.

File: oap-shuffle/remote-shuffle/README.md (hunk 2 above)
Comment (gczsjdy) [reply oap-shuffle/remote-shuffle/README.md]: Right!

**Comment by gczsjdy** [comment]
Remote Shuffle was originally developed at https://github.com/Intel-bigdata/RemoteShuffle, it's merged to OAP as a chunk commit. Please refer to the original repository for the complete evolving history.

**Pull request merged**